src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -759,7 +759,146 @@ private ITypeReference GetUnionTypeType(UnionTypeSyntax syntax)
+                            DiagnosticBuilder.ForPosition(memberSyntax)
+                                .DiscriminatedUnionMemberAdditionalPropertiesUnsupported());
+
+                        continue;
+                    }
+
+                    // validate the discriminator property value does not overlap with other members
+                    var discriminatorMemberLiteral = discriminatorTypeProperty.TypeReference.Type as StringLiteralType;
+                    var discriminatorMemberValue = discriminatorMemberLiteral!.RawStringValue;
+
+                    if (memberDiscriminatorValues.Contains(discriminatorMemberValue))
+                    {
+                        AddErrorDiagnostic(DiagnosticBuilder.ForPosition(memberSyntax)
+                            .DiscriminatorPropertyMemberDuplicatedValue(discriminatorPropertyName, discriminatorMemberValue));
+
+                        continue;
+                    }
+
+                    memberDiscriminatorValues.Add(discriminatorMemberValue);
+                    expandedMemberTypes.Add(objectType);
+                }
+                else if (memberTypeEvaluated is DiscriminatedObjectType memberDiscriminatedObjectType)
+                {
+                    // validate it has the same discriminator property
+                    if (memberDiscriminatedObjectType.DiscriminatorProperty.Name != discriminatorPropertyName)
+                    {
+                        AddErrorDiagnostic(DiagnosticBuilder.ForPosition(memberSyntax).DiscriminatorPropertyNameMustMatch(discriminatorPropertyName));
+
+                        continue;
+                    }
+
+                    // validate there's not value overlap
+                    foreach (var nestedDiscriminatorValue in memberDiscriminatedObjectType.UnionMembersByKey.Keys)
+                    {
+                        if (memberDiscriminatorValues.Contains(nestedDiscriminatorValue))
+                        {
+                            AddErrorDiagnostic(DiagnosticBuilder.ForPosition(memberSyntax)
+                                .DiscriminatorPropertyMemberDuplicatedValue(discriminatorPropertyName, nestedDiscriminatorValue));
+                        }
+
+                        memberDiscriminatorValues.Add(nestedDiscriminatorValue);
+                    }
+
+                    expandedMemberTypes.AddRange(memberDiscriminatedObjectType.UnionMembersByKey.Values);
+                }
+                else
+                {
+                    return ErrorType.Create(
+                        DiagnosticBuilder.ForPosition(discriminatorDecorator)
+                            .DiscriminatorDecoratorOnlySupportedForObjectUnions());
+                }
+            }
+
+            if (errorDiagnostics?.Any() ?? false)
+            {
+                return ErrorType.Create(errorDiagnostics);
+            }
+
+            return new DiscriminatedObjectType(
+                TypeHelper.GetNormalizedTypeListUnionName(expandedMemberTypes),
```

Comment:
[nit]: `expandedMemberTypes` is already a flattened list of `ObjectType`s; couldn't you just call `string.Join(" | ", expandedMemberTypes.Select(t => t.Name))` instead of passing the list to a recursive flattener?

Reply:
[dd5cea7](https://github.com/Azure/bicep/pull/11222/commits/dd5cea7ec62d7b867977e6ea63253b13da2e90d4)

Updated. I added a method to `TypeHelper` to get the sorted type names using the comparer there as they were appearing in random order.